### ISSUE TITLE
Grading incorrectly discarded a winning OPR in block 224993

### ISSUE THREAD

**Issue by PaulBernier** [pegnet/pegnet#343]
Today, the Orax mining pool forked from the rest of the miners at block 224993. The Orax pool has its own completely separate implementation of PegNet grading algo. After a long deep dive in what happened I believe the Orax implementation was the correct one. Obviously as the majority always win, the Orax pool yielded and joined back the majority of miners now. Here's what I found:

In block 224993 those 2 entries were submitted:
- https://explorer.factoid.org/entry?hash=57979cb272da64502e4f4bb219d5177ca04631d66e81257fdc4384359895e764
- https://explorer.factoid.org/entry?hash=f5b7a93b2824383e4963d6b3132fd071722d4e9088542333b710d5d27a344047
You will notice they are almost identical: same nonce, same version, same entry content. The only difference is the self reported difficulty. `57979cb2` self reported difficulty is wrong, and `f5b7a93b` self reported difficulty is correct. `57979cb2` appears in the entry block before `f5b7a93b`. 
Core implementation of grading perform an operation to filter duplicates: https://github.com/pegnet/pegnet/blob/master/modules/grader/baseGradedBlock.go#L79 The issue is that this de-duplication happen _before_ the OPR is fully validated. In our case here `57979cb2` was kept and `f5b7a93b` discarded, even though `57979cb2` is not even a valid OPR. But here this biggest problem is that `f5b7a93b` was am actual winning OPR in terms of difficulty and grading. Orax rightfully gave `f5b7a93b` its place in the top25, while core implementation didn't, resulting in a fork.

TL;DR: `f5b7a93b` that was a perfectly valid winning OPR was incorrectly discarded, yielding wrong set of winners. 

I think that this deduplication logic doesn't make much sense in the first place, because if an OPR is valid, then you cannot have duplicate nonce and hash, otherwise the entries would be strictly identical and couldn't be included into Factom in the first place. If you wish to keep it's fine, but then we need to validate the self reported difficulty before. Thoughts? @Emyrk @WhoSoup

**Comment by Emyrk** [comment]
This is a good find. I only looked at this briefly, but I believe you are correct. I'm traveling all tomorrow, thanks for bringing this up

**Comment by Emyrk** [comment]
This also happened at 225337

**Comment by WhoSoup** [comment]
While this discrepancy between pegnet and pegnetd has been addressed in https://github.com/pegnet/pegnet/pull/345, it does raise a very interesting issue: Why did MiningCenter submit two OPRs with different self reported difficulties? Both of the OPRs linked in paul's post are paid for by the same EC address, so this is not an issue of a malicious third party trying to bump out an OPR.

Does anyone know if they're running a custom mining implementation?

**Comment by Emyrk** [comment]
@WhoSoup They are running a custom mining implementation based off of the Reference miner. I had reached out to him before when this was happening more frequently. He thanked me and I thought he fixed his bug.

From what he told me, he is not a GoLang dev, so it might have been a bit of a hacky job.
